modules/imap/hm-imap.php

```diff
@@ -2160,6 +2160,16 @@ public function get_mailbox_page($mailbox, $sort, $rev, $filter, $offset=0, $lim
             if ($keyword) {
                 $uids = $this->search($filter, $uids, array(array('TEXT', $keyword)));
             }
+            if ($trusted_senders && is_array($trusted_senders)) {
+                $terms = array();
+                if ($keyword) {
+                    $terms[] = array('TEXT', $keyword);
+                }
+                foreach ($trusted_senders as $sender) {
+                    $terms[] = array('FROM', 'NOT '. $sender);
+                }
+                $uids = $this->search($filter, $uids, $terms);
```

Comment:
There should be only one search call - please combine this block with the keyword block above and execute one final search with the relevant terms.